readers/apiserver/watchlist/internal/watchlist/relist/listers.go

```diff
@@ -0,0 +1,136 @@
+package relist
+
+import (
+	"context"
+	"fmt"
+	"reflect"
+
+	"github.com/Azure/tattler/data"
+	"github.com/gostdlib/base/retry/exponential"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+)
+
+var backoff = exponential.Must(exponential.New())
+
+// genericLister is a function that lists resources of type T. It can be any of the common
+// client list functions that return a concrete type, such as *corev1.NamespaceList.
+// So for example, r.clientset.CoreV1().Namespaces().List is a genericLister[*corev1.NamespaceList].
+type genericLister[T runtime.Object] func(ctx context.Context, options metav1.ListOptions) (T, error)
+
+// adapter will turn any genericLister into a ListPage. A generic lister is any of the common client
+// list functions that return a concrete type, such as *corev1.NamespaceList. The returned ListPage
+// also will handle exponential backoff retries for retriable errors.
+func adapter[R runtime.Object](f genericLister[R]) listPage {
+	return func(ctx context.Context, options metav1.ListOptions) ([]data.Entry, string, error) {
+		var entries []data.Entry
+		var continueToken string
+		err := backoff.Retry(
+			ctx,
+			func(ctx context.Context, _ exponential.Record) error {
+				obj, err := f(ctx, options)
+				if err != nil {
+					return err
+				}
+				entries, err = listToEntries(obj)
+				if err != nil {
+					return err
+				}
+				continueToken, err = getContinue(obj)
+				if err != nil {
+					return err
+				}
+				return nil
+			},
+			exponential.WithMaxAttempts(10),
+		)
+		return entries, continueToken, err
+	}
+}
+
+// listToEntries converts a List() result to a slice of data.Entry.
+func listToEntries(listResult runtime.Object) ([]data.Entry, error) {
+	var events = make([]data.Entry, 0, 100)
+
+	if listResult == nil {
+		return nil, fmt.Errorf("listToEntries: listResult is nil: %w", exponential.ErrPermanent)
+	}
+
+	// Use reflection to extract Items field from any List type (PodList, NodeList, etc.)
+	listValue := reflect.ValueOf(listResult)
+	if listValue.Kind() == reflect.Ptr {
+		listValue = listValue.Elem()
+	}
+
+	itemsField := listValue.FieldByName("Items")
+	if !itemsField.IsValid() {
+		return nil, fmt.Errorf("listToEntries: no Items field found in %T: %w", listResult, exponential.ErrPermanent)
+	}
+
+	for i := 0; i < itemsField.Len(); i++ {
+		item := itemsField.Index(i)
+		if item.CanInterface() {
+			// Get the interface value
+			itemInterface := item.Interface()
+
+			// Try as runtime.Object directly first (for pointer types)
+			if obj, ok := itemInterface.(runtime.Object); ok {
+				events = append(events, data.MustNewEntry(obj, data.STWatchList, data.CTSnapshot))
+				continue
+			}
+
+			// For value types, we need to take their address
+			if item.CanAddr() {
+				itemPtr := item.Addr()
+				if itemPtr.CanInterface() {
```

Comment:
seems like this could be if ! caninterface, return nil. Then you have one less `else`.

Reply:
Done